File: 6M4W/ehowlqk/boj_1389.js
function findKevinBacon(i, edges) {
  const visited = Array(n).fill(10000);
  visited[i] = 0;
  const kevinBacon = dfs(i + 1, 0, edges, visited);
  return kevinBacon.reduce((a, b) => a + b);
}

function dfs(member, depth, edges, visited) {
  edges[member - 1].forEach((friend) => {
    if (visited[friend - 1] > depth + 1) {
      visited[friend - 1] = depth + 1;
      visited = dfs(friend, depth + 1, edges, visited);
    }
  });

  return visited;
}

const fs = require("fs");
const input = fs
  .readFileSync("./6M4W/ehowlqk/boj_1389.txt")
  .toString()
  .trim()
  .split("\n");

const [n, m] = input[0].split(" ").map(Number);

const edges = Array.from({ length: n }, () => []);

// 인접 배열
for (let i = 1; i <= m; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  edges[a - 1].push(b);
  edges[b - 1].push(a);
}

let answer = 0;
let minKevinBacon = 1000;

for (let i = 0; i < n; i++) {
  const kevinBacon = findKevinBacon(i, edges);
  if (minKevinBacon > kevinBacon) {
    answer = i + 1;
    minKevinBacon = kevinBacon;
  }
}

console.log(answer);

File: 6M4W/ehowlqk/boj_1389.py
n, m = map(int, input().split())

# 거리 배열 초기화
distances = [[0 if row == col else 1e9 for col in range(n)] for row in range(n)]

# 노드 사이 거리 입력
for _ in range(m):
    a, b = map(int, input().split())
    distances[a-1][b-1] = 1
    distances[b-1][a-1] = 1

# 노드 간 거리 업데이트
for k in range(n):
    for i in range(n):
        for j in range(n):
            distances[i][j] = min(distances[i][k] + distances[k][j], distances[i][j])

answer = 0
minKevinBacon = 1e9

# 케빈 베이컨 가장 작은 유저 구하기
for user in range(1, n+1):
    kevinBacon = sum(distances[user - 1])
    if kevinBacon < minKevinBacon:
        minKevinBacon = kevinBacon
        answer = user

print(answer)
